### DISCUSSION TITLE
Is there a way to disconnect from redis after using PubSub?

### DISCUSSION THREAD

**Discussion by pangxie100** [ra1u/redis-dart#64] [Q&A]
Hi, 
I really appreciate your work and find it easy to use. When I tried to use PubSub, I found a description said "Once PubSub is created, Command is invalidated and should not be used on the same connection." If we use PubSub, is there any way to disconnect from Redis since cmd.send_object(["QUIT"]) doesn't work any more?

Thank you.

**Comment by ra1u** [comment]
Why we don't (currently) allow using socket for both send_object and PubSub.

Library is designed around assumption that there is exactly one response for every request, For this reason we can not correctly distinguish received response from received  sub message. For example, there could arrive subscription message on socket, just after request was sent and before response was received on same socket. We do not dispatch messages by type and values but only by order and we dispatch them in same order as they were sent. For similar reason it is not a option to have more than one subscriber on single connection.

What I suggest is that you create new and separate `RedisConnection` for receiving subscription messages. I hope that this make sense.

About close, you can just close connection:
In `RedisConnection` there is [`close()`](https://github.com/ra1u/redis-dart/blob/6d22c6a66b54a6572f8ca7ec1ea2a7bfa618cfa2/lib/connection.dart#L47-L51) and in `Command` (what you usually have) there is [`get_connection()`](https://github.com/ra1u/redis-dart/blob/6d22c6a66b54a6572f8ca7ec1ea2a7bfa618cfa2/lib/command.dart#L84-L86)
So that  `cmd.get_connection().close()` should close socket.

Kind regards, Luka

**Reply by pangxie100** [reply]
Thank you for replying with a great explanation! 

I think close() function is exactly what I want to close the connection. 

However, when I use it after unsubscribing all the channels, there is another error on Flutter called "complete error" with "[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: stream is closed". The possible reason is that I didn't do anything to cancel the effect of pubsub.getStream().listen(). So when the connection is closed, it still tries to listen to something.  

Is there any way to stop listening to the stream? Or this is an inevitable error each time when we close a connection on Pubsub part?

Thank you.

**Reply by ra1u** [reply]
I see.
Can please you provide some some code that can reproduce this issue and can speed up finding appropriate solution?

Kind regards, Luka.

**Reply by ra1u** [reply]
I think that is suspect where issue might be and that using closed connection during pubsub results in exception that is not handled appropriately.

https://github.com/ra1u/redis-dart/blob/6d22c6a66b54a6572f8ca7ec1ea2a7bfa618cfa2/lib/pubsub.dart#L18-L22

This seems like a bug.

**Comment by ra1u** [comment]
@pangxie100 on current master we have now commit fc7c3bce7d4c2c60b408f604fd723b3eaa5e67e9 that should improve on this issue.
Improvement is, that you can call close on PubSub connection as described before and you can also catch error on subscribed stream.

~~~~ Dart
pubsub.getStream().listen((message){
  print("message: $message");
},
onError : (e){
  print("error: $e");
});
~~~~

Let me know if that works for you and so we can roll out new release .


Kind regards, Luka.

**Reply by ra1u** [reply]
This is now released in v3.1.0 . Additionally there were some improvements since [previous post](https://github.com/ra1u/redis-dart/discussions/64#discussioncomment-2623753). Please take a look at [updated example](https://github.com/ra1u/redis-dart/#pubsub) . Let me know if you find and issue so we can fix such issue.
